### PR TITLE
Add training script, config class, and gtr runner

### PR DESCRIPTION
**Pull request by aaprasad**
add stuff to gitignore
write config class to handle config parsing and loading for training
add sleap + microscopy datasets
add train script
add tracker for inference with postprocessing
add gtr_runner to handle training, inference, and validation logic

### CODE REVIEW

File: biogtr/config.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+import pytorch_lightning as pl
+from biogtr.models.global_tracking_transformer import GlobalTrackingTransformer
+from biogtr.training.losses import AssoLoss
+from biogtr.datasets.sleap_dataset import SleapDataset
+from biogtr.datasets.microscopy_dataset import MicroscopyDataset
+from omegaconf import DictConfig, OmegaConf
+from typing import Union, Iterable
+
+
+class Config:
```
Comment (sheridana) [inline biogtr/config.py:16]: We should have a test for the config getters/setters. However, having codecov target at 100% is probably overkill. We should maybe have a threshold of 20% (will add a codecov.yml to deal with this.)

File: biogtr/datasets/data_utils.py [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+
+    return sorted_anchors
+
+
+"""
+PARSERS
+"""
+
+
+def parse_trackmate_xml(xml_path: str) -> pd.DataFrame:
```
Comment (sheridana) [inline biogtr/datasets/data_utils.py:195]: We should consolidate the parsing logic (I can handle this) and add appropriate tests.

File: biogtr/datasets/sleap_dataset.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            crop_size: the size of the object crops
+            chunk: whether or not to chunk the dataset into batches
+            clip_length: the number of frames in each chunk
+            crop_type: `centroid` or `pose` - determines whether to crop around a centroid or around a pose
+            mode: `train` or `val`. Determines whether this dataset is used for training or validation.
+            Currently doesn't affect dataset logic
+            tfm: The augmentation function from albumentations
+            tfm_cfg: The config for the augmentations
+        """
+        self.slp_files = slp_files
```
Comment (sheridana) [inline biogtr/datasets/sleap_dataset.py:41]: Should we handle directories here or at high level?

File: biogtr/datasets/tracking_dataset.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+from biogtr.datasets.sleap_dataset import SleapDataset
+
+# todo: move to config
+num_workers = 0
+shuffle = True
+
+device = "cuda" if torch.cuda.is_available() else "cpu"
+
+if num_workers > 0:
+    # prevent too many open files error
```
Comment (sheridana) [inline biogtr/datasets/tracking_dataset.py:16]: This logic should be handled at high level entry point - I believe this is fixed in subsequent PRs but please confirm.

File: biogtr/inference/metrics.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    Args:
+        switches: a dict of dicts containing the mislabeled trajectories
+        and the frames at which they occur
+
+    Returns:
+        the number of switched labels in the video chunk
+    """
+    only_switches = {k: v for k, v in switches.items() if v != {}}
+    sw_cnt = sum([len(i) for i in list(only_switches.values())])
+    return sw_cnt
```
Comment (sheridana) [inline biogtr/inference/metrics.py:88]: I think we decided on pymot for eval (seems to be cleanest, but maybe we should benchmark with trackeval for our use?). Either way, this functionality should be added to this file.

File: biogtr/inference/post_processing.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+    The box order must be (xmin, ymin, xmax, ymax)
+
+    Args:
+        boxes1,boxes2 (Boxes): two `Boxes`. Contains N & M boxes, respectively.
+
+    Returns:
+        Tensor: intersection, sized [N,M].
+    """
+    boxes1, boxes2 = boxes1.tensor, boxes2.tensor
```
Comment (sheridana) [inline biogtr/inference/post_processing.py:56]: back to previous boxes comment - could probably just remove Boxes class and take tensor input directly (just handle shapes etc when calling). This can be a downstream change.

File: biogtr/inference/tracker.py [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+from scipy.optimize import linear_sum_assignment
+from copy import deepcopy
+
+
+class Tracker:
+    """Tracker class used for assignment based on sliding inference from GTR."""
+
+    def __init__(
+        self,
+        model: GlobalTrackingTransformer,
```
Comment (sheridana) [inline biogtr/inference/tracker.py:17]: We discussed this - maybe better for tracker to take asso preds as input and keep the model prediction separate. Can deal with this later, fine for now.

File: biogtr/models/attention_head.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
+"""Module containing different components of multi-head attention heads."""
+
+import torch
+import torch.nn.functional as F
+
+# todo: add named tensors
+
+
+class MLP(torch.nn.Module):
```
Comment (sheridana) [inline biogtr/models/attention_head.py:9]: These classes are also adapted from gtr repo so we should just make sure to cite.

File: biogtr/models/global_tracking_transformer.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
+"""Module containing GTR model used for training."""
+from biogtr.models.transformer import Transformer
+from biogtr.models.visual_encoder import VisualEncoder
+from torch import nn
+
+# todo: do we want to handle params with configs already here?
+
+
+class GlobalTrackingTransformer(nn.Module):
```
Comment (sheridana) [inline biogtr/models/global_tracking_transformer.py:9]: Do we also handle high level instantiation via hydra? maybe this is already added in subsequent PRs

File: biogtr/models/gtr_runner.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+        if "sw_cnt" in eval_metrics:
+            tracker = Tracker(self.model, **self.tracker_cfg)
+            instances_pred = tracker(instances)
+            matches, indices, _ = metrics.get_matches(instances_pred)
+            switches = metrics.get_switches(matches, indices)
+            sw_cnt = metrics.get_switch_count(switches)
+            return_metrics["sw_cnt"] = sw_cnt
+        return return_metrics
+
+    def configure_optimizers(self) -> dict:
```
Comment (sheridana) [inline biogtr/models/gtr_runner.py:160]: Add functions to init optimizers and schedulers (i think this is added in subsequent PRs, please confirm).

File: biogtr/models/model_utils.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    Returns:
+        asso_output: Exactly the same as before but with the softmax applied.
+        # ------------------------ An example of asso_output ----------------------- #
+        N_i: number of detected instances in i-th frame of window.
+        N_t: number of instances in current/query frame (rightmost frame of the window).
+        T: length of window.
+        asso_output is of shape: (T, N_t, N_i).
+    """
+    # N_i: number of detected instances in i-th frame of window.
```
Comment (sheridana) [inline biogtr/models/model_utils.py:48]: can remove these comments since they are in the docstrings

File: biogtr/models/model_utils.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    # N_t: number of instances in current frame (rightmost frame of the window).
+    # T: length of window.
+
+    # asso_output: (T, N_t, N_i)
+
+    asso_active = []
+    for asso in asso_output:
+        # asso: (N_t, N_i)
+
+        # I'm guessing what they are doing here is giving the model a chance to pick "uncertain".
```
Comment (sheridana) [inline biogtr/models/model_utils.py:58]: can remove this comment. this function is just to get probs from logits during inference.

File: biogtr/training/train.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+    Args:
+        cfg: The config dict parsed by `hydra`
+    """
+    train_cfg = Config(cfg)
+
+    # update with extra cli args
+    hparams = {}
+    for arg in sys.argv[1:]:
+        if arg.startswith("+"):
```
Comment (sheridana) [inline biogtr/training/train.py:39]: think we already discussed handling this better (e.g directly with hydra)?

**Review by sheridana — COMMENTED** [review]
I think generally speaking a lot of the raised issues are addressed in subsequent PRs that we already discussed. We should handle codecov to have some flexibility. Small comments added for confirmation, minor things to address then should be good to merge.

File: biogtr/config.py (hunk 1 above)
Comment (aaprasad) [reply biogtr/config.py:16]: ok will add tests for config

File: biogtr/datasets/tracking_dataset.py (hunk 4 above)
Comment (aaprasad) [reply biogtr/datasets/tracking_dataset.py:16]: Need to remove it from this file

File: biogtr/inference/boxes.py [hunk 14]
```diff
@@ -0,0 +1,214 @@
+"""Module containing Boxes class."""
+from typing import List, Tuple, Union
+import torch
+
+
+class Boxes:
```
[the remark this answers is not in the document]
Comment (aaprasad) [reply biogtr/inference/boxes.py:6]: Yeah its only used in pairwise iou

File: biogtr/models/global_tracking_transformer.py (hunk 9 above)
Comment (aaprasad) [reply biogtr/models/global_tracking_transformer.py:9]: Yeah the GTR is instantiated within GTRRunner which is handled by the hydra config

File: biogtr/models/gtr_runner.py (hunk 10 above)
Comment (aaprasad) [reply biogtr/models/gtr_runner.py:160]: Yep see [model_utils](https://github.com/talmolab/biogtr/blob/d4ff53d5d237f3c183c3e4341617c2523fbb2c99/biogtr/models/model_utils.py#L71)

File: biogtr/training/train.py (hunk 13 above)
Comment (aaprasad) [reply biogtr/training/train.py:39]: Yep see[here](https://github.com/talmolab/biogtr/blob/faa309bb012bc2e927f8db98ebba266ed07383cc/biogtr/training/train.py#L42)

File: biogtr/datasets/sleap_dataset.py (hunk 3 above)
Comment (aaprasad) [reply biogtr/datasets/sleap_dataset.py:41]: assume handling at high level in train.py/inference.py etc

File: biogtr/inference/boxes.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (aaprasad) [reply biogtr/inference/boxes.py:6]: TODO downstream
